lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -1255,9 +1252,16 @@ matchWitness(TypeChecker &tc,
   };
 
   // Match a type in the requirement to a type in the witness.
-  auto matchTypes = [&](Type reqType, Type witnessType) 
+  auto matchTypes = [&](Type reqType, Type witnessType)
                       -> Optional<RequirementMatch> {
-    cs->addConstraint(ConstraintKind::Equal, reqType, witnessType, locator);
+    ConstraintKind kind = ConstraintKind::Conversion;
+    if (auto func = dyn_cast<FuncDecl>(req)) {
+      auto name = req->getFullName().getBaseName();
+      if (func->isOperator() && name.str() == "==") {
+        kind = ConstraintKind::Equal;
+      }
+    }
+    cs->addConstraint(kind, witnessType, reqType, locator);
```

Comment:
:-/ Why? I thought to get covariant behaviour we should always match witness agains requirement.

Also, `matchFunctionTypes` flips function arguments itself, as comment states:
> // Input types can be contravariant (or equal)

but can they? Seems there is some inconsistency:

```swift
func subtypeArgument1(_ x: (_ fn: ((String) -> Int)) -> Int) { }
func testSubtypeArgument1(x2: (_ fn: ((String) throws -> Int)) -> Int) {
  subtypeArgument1(x2)
}
```
This works fine.

```swift
func subtypeArgument1(_ x:(Int) -> Int) {}
func testSubtypeArgument1(x2: (Int) throws -> Int) {
  subtypeArgument1(x2)
}
```
This gives an error:
> error: invalid conversion from throwing function of type '(Int) throws -> Int' to non-throwing function type '(Int) -> Int'

Is this behavior intended and should we persist it?

That's why I've used `locator` to disable that flipping for witness function arguments and always match witness agains requirement.